### PR TITLE
More robust/thread safe test server startup sequence.

### PR DESCRIPTION
**Pull request by ioquatix**
Upon reviewing the test code, I'm not convinced the startup "join" is really that great. This uses a `Thread::Queue` to wait until the server starts up successfully (or crashes).

Potential fix for https://github.com/rack/rack/issues/1859 ?

### CODE REVIEW

**Comment by ioquatix** [comment]
Okay, this doesn't seem to solve the problem.

**Comment by simi** [comment]
Is this problem only on JRuby? Maybe we can ask JRuby folks to help us in here. I was looking at this problem as well, but I wasn't able to reproduce it locally.

**Comment by ioquatix** [comment]
I've asked JRuby and TruffleRuby folks for any advice.

The code looks less than ideal. It's running a server in the same process and killing it by sending itself SIGINT.

My guess is something isn't playing nice. Either thread bug or race condition, but it's impossible to tell right now. I'll mess around on this PR a bit more.

**Comment by ioquatix** [comment]
I also wonder if:

```
      Port: TCPServer.open('localhost', 0){|s| s.addr[1] },
```

might be racy.

File: test/spec_server.rb [hunk 1]
```diff
@@ -334,6 +334,28 @@ def start(*) [self.class, :started] end
...
+      thread = Thread.new do
+        begin
+          server.start { |s| ready.push(true) }
+        ensure
+          ready.close
+        end
+      end
+
+      unless ready.pop
+        thread.join
```
Comment (eregon) [inline test/spec_server.rb:349]: This seems very hard to follow, what is it trying to do?
Can't it just always join?

Comment (ioquatix) [reply test/spec_server.rb:349]: It only joins if it fails to start the server.

File: test/spec_server.rb [hunk 2]
```diff
@@ -334,6 +334,28 @@ def start(*) [self.class, :started] end
...
+      end
+
+      sleep 10
```
Comment (eregon) [inline test/spec_server.rb:352]: This seems hacky, sleep forever or not sleep seems more reliable

Comment (ioquatix) [reply test/spec_server.rb:352]: It's a total hack to see if it's a race condition or something else.

File: test/spec_server.rb [hunk 3]
```diff
@@ -349,8 +371,9 @@ def start(*) [self.class, :started] end
       daemonize: false,
       server: 'webrick'
     )
-    t = Thread.new { server.start { |s| Thread.current[:server] = s } }
-    t.join(0.01) until t[:server] && t[:server].status != :Stop
+
+    t = start_server(server)
+
     body = if URI.respond_to?(:open)
```
Comment (eregon) [inline test/spec_server.rb:377]: Why not have all that logic in the `server.start` block?
Is the server stopped when the block ends?

Comment (ioquatix) [reply test/spec_server.rb:377]: I have no idea about the deeper implementation, I just tried to move the existing shared logic into a method.

**Comment by eregon** [comment]
> I also wonder if:
> 
> ```
>       Port: TCPServer.open('localhost', 0){|s| s.addr[1] },
> ```
> 
> might be racy.

Yes, it is. Can we rather pass `0`/`nil` for the Port and that then uses whatever is available?

**Comment by ioquatix** [comment]
> Yes, it is. Can we rather pass 0/nil for the Port and that then uses whatever is available?

I think the problem with that is we don't know the port to connect to... well I guess we might? But there is no interface for passing it back AFAIK.

**Comment by eregon** [comment]
The `Process.kill(:INT, $$)` seems fragile indeed, that basically raises an Interrupt on the main thread (unless a custom handler is set with #trap).
Maybe the test just wants to t.kill or t.raise?

**Comment by ioquatix** [comment]
> Maybe the test just wants to t.kill or t.raise?

Yes I think so.

**Comment by eregon** [comment]
> I think the problem with that is we don't know the port to connect to... well I guess we might? But there is no interface for passing it back AFAIK.

Should the block given to start indicate the socket is opened, and then one can get the port from socket.addr[1]?

**Comment by ioquatix** [comment]
> Shouldn't the block given to start indicate the socket is opened, and then one can get the port from socket.addr[1]?

Maybe?

**Comment by eregon** [comment]
The `Process.kill(:INT, $$)` is likely to trigger the custom SIGINT handler, if Rack::Server somehow sets one up?

**Comment by ioquatix** [comment]
Does `trap(:int)` work with `Thread#raise(Interrupt)`?

https://github.com/rack/rack/blob/main/lib/rack/server.rb#L331-L337

**Comment by eregon** [comment]
No, it's only if the signal is sent.
So this could be a bit more direct with calling `server.shutdown` directly.

**Comment by ioquatix** [comment]
https://github.com/rack/rack/blob/56fe8ae3d187513aca3ba4bd905bdca2dba42d01/lib/rack/server.rb#L339

I don't think we can get any sockets out of this generally speaking.

**Comment by simi** [comment]
If the problem is only within this test, wouldn't be easier to just write small bash script to start the server in one process, wait unless it is up and test request using some CLI tool? We can run it as a part of CI.

**Comment by eregon** [comment]
Does `server.start { ... }` block until it processes all requests/forever (after it started and called the block)?
Seems very confusing :p

File: test/spec_server.rb (hunk 1 above)
Comment (eregon) [reply test/spec_server.rb:349]: Another way would be to set `abort_on_exception` for that thread as the first thing it does. Then no need for the queue.

**Comment by ioquatix** [comment]
Not just confusing, potentially broken... anyway, here is the webrick implementation which is what is used by this test:

https://github.com/rack/rack/blob/56fe8ae3d187513aca3ba4bd905bdca2dba42d01/lib/rack/handler/webrick.rb#L29-L45

**Comment by eregon** [comment]
Does anything use Rack::Server? Seems pretty unusable.

Anyway, I'm out of ideas, but from the failing backtrace on JRuby, it sounds like it could be a problem of net/http or of that spec_server being incorrect.
FWIW I wrote a small HTTP server for testing in https://github.com/ruby/spec/blob/master/library/net/http/http/fixtures/http_server.rb, there were quite a few subtleties.

**Comment by eregon** [comment]
Ah and WEBRick's shutdown is fundamentally racy and broken, so that may be part of it too, and that's why I wrote that HTTP server, WEBrick isn't reliable enough for a clean shutdown.

**Comment by ioquatix** [comment]
I see so we might end up with the previous webrick server corrupting the process?

**Comment by ioquatix** [comment]
Is it possible to fix WEBRick or is the problem too deeply ingrained into the design/public interface?

**Comment by ioquatix** [comment]
Now I see TruffleRuby fail: https://github.com/rack/rack/runs/5884224992?check_suite_focus=true#step:8:178

image

**Comment by simi** [comment]
@ioquatix I have seen this error when I was setting 2-2-stable CI as well on TruffleRuby. Seems random as well.

**Comment by eregon** [comment]
That error just looks like a typical race between the exists check and the File.delete/unlink.

Re WEBrick I'd think it would be fixable, but the shutdown pipe design would need to be rethought carefully to avoid concurrent write_nonblock+close IIRC.

**Comment by simi** [comment]
What about to skip this spec at JRuby for now?

**Comment by ioquatix** [comment]
We could do... but @headius said he'd take a look on Monday.

**Review by Newman1515 — COMMENTED** [review]
Got it

**Comment by ioquatix** [comment]
We are going to punt on this, probably by extracting `rackup` to a separate gem.
